### DISCUSSION TITLE
`database is locked` error when use sqlite db

### DISCUSSION THREAD

**Discussion by bluzy** [apache/airflow#33199] [General] [answered]
### Apache Airflow version

2.6.2

### What happened

I am testing dags on local airflow with default sqlite db.

After upgrade airflow version to 2.6.2 from 2.2.4, I encountered `database is locked` error when run `airflow db init` or `airflow standalone` command.

I inspected code, then I found that session is not released after delete `serialized_dag` table.

https://github.com/apache/airflow/blob/5a0494f83e8ad0e5cbf0d3dcad3022a3ea89d789/airflow/utils/db.py#L886-L893

After I added `session.commit()` after deleting `serialized_dag` table, it works fine.
Or, manually delete rows in `serialized_dag` table fix this issue..

The problem occurs with our production dag codes, not reproduced with simple example dag codes.
I think only `data` fields' size are difference, but I don't know why it makes db lock..

And, also here has same db lock problem.

https://github.com/apache/airflow/blob/624cf7f75a95de63bcf7a22016b13f917669e9f2/airflow/utils/db.py#L1022-L1043

Likewise, Adding `session.commit()` after `sqlalchemy.schema.MetaData()` solves problem.

### How to reproduce

Only reproduced with sqlite db.
I don't know exact condition, but I guess large dags cause the problem..

### Operating System

MacOS (Apple Silicon)

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment, marked as answer]
First of all `sqlitedb` should only be used for local unit testing and quick start (see the documetation where we clearly warn against using it for anything else)

Secondly - the _reserilalize_dag is only used in "upgradedb" that is closing session (via @provide_session decorator - and it's precisely as plannec to not close sessons.

One of the main reasons of "development" status is that Sqlite cannot work when you have two concurrent accesses to the sqlite file, this creates a lock - which is likely what you observed. you likley had another airflow process still running and accessing the database while you were running upgrade - this is likely why you saw the lock happen. Don't do it. Don't use sqlite for anything else than local unit testing and quick start, you will have all kinds of similar problems like that what you do.

Converted to discussion in case more of it is needed.

**Comment by DFINITYManu** [comment]
This is definitely a bug in the `reflect_tables()` code.  I'm using 2.9.1 so the bug clearly has not been fixed from 2.6.x to 2.9.x.  

And those of us using Airflow during development really do not appreciate being told "don't use it for production" -- yes, we know that already.

**Reply by DFINITYManu** [reply]
Actually, the bug hasn't even been fixed in `main`.  I still get this issue, and this workaround still works flawlessly.

**Reply by potiuk** [reply]
If you have an idea how to fix it - feel free. This discussion is from 2023 - which is 2 years ago. Since then not only a lot of things happened, but also we changed the mode of using sqlite - to actually allow concurrency (which was not there 2 years ago)l

So if you have anything to say that changes it here and if there is a **similar** issue in main - the best way to do something about it is to create a PR fixing it @DFINITYManu . The second best, is to create an issue describing in detail what problem you have with the "current" main - including stack traces and what kind of workaround helps. Commenting on 2 years old discussion with vague comment that might or might not have to do anything with the issue you see currently in main does not move a needle with solving the issue.

**Reply by potiuk** [reply]
And yes - 3 years ago  and even for any Airflow 2 version my answer was "DO NOT USE it in multi-process access" because it has never been designed to do it. 

With Airflow 3 - i..e. main, this is a different case because ... we added that capability ... so if you see any issue in `main` about it - PLEASE open an issue rather than complain about 2 years old comment, Venting your frustration by reading 2 years old post does not really help with solving the issue.
